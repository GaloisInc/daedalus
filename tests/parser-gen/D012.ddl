-- This is to test a subfragment of plain-ppm.ddl
-- in particular the fixed lower bound on Many


def Token P = {
  $$ = P;
  Many (1..) WS;
}

def RGB = {
  red   = Token Natural;
  green = Token Natural;
  blue  = Token Natural;
}



-- def WS = $[0 | 9 | 12 | 32 | '\n' | '\r']
def WS = $[0]


def Natural = {
  @ds = Many (1..) Digit;
  ^ for (val = 0; d in ds) (addDigit val d);
}

def Digit = {
  @d = $['0' .. '9'];
  ^ d - '0';
}


def addDigit val d =
  10 * val + (d as int)

def Main =
  { RGB
  ; END
  }
