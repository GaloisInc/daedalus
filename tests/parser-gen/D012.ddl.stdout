module D012
 
--- Imports:
 
--- Type defs:
type D012.RGB = { red: int
                ; green: int
                ; blue: int
                }
 
--- Rules:
 
D012._WS : Grammar {} =
  @Match {0}
 
D012.Token ?a0 (P : Grammar ?a0) : Grammar ?a0 =
  do ($$ : ?a0) <- P
     @Many[ 1 .. ] D012._WS
     pure $$
 
D012.Digit : Grammar (uint 8) =
  do (d : uint 8) <- Match ('0' .. '9')
     ($$ : uint 8) <- pure (d - '0')
     pure $$
 
D012.addDigit ?a0 (Coerce safe ?a0 int) (val : int) (d : ?a0) : int =
  (10 * val) + (d as int)
 
D012.Natural : Grammar int =
  do (ds : [uint 8]) <- Many[ 1 .. ] D012.Digit
     ($$ : int) <- pure (for ((val : int) = 0; (d : uint 8) in ds)
                           (D012.addDigit (uint 8) val d))
     pure $$
 
D012.RGB : Grammar D012.RGB =
  do (red : int) <- D012.Token int D012.Natural
     (green : int) <- D012.Token int D012.Natural
     (blue : int) <- D012.Token int D012.Natural
     pure {red = red,
           green = green,
           blue = blue}
 
D012.WS : Grammar (uint 8) =
  Match {0}
 
D012._Token ?a0 (_P : Grammar {}) : Grammar {} =
  do _P
     @Many[ 1 .. ] D012._WS
 
D012._Digit : Grammar {} =
  @Match ('0' .. '9')
 
D012._Natural : Grammar {} =
  @Many[ 1 .. ] D012._Digit
 
D012._RGB : Grammar {} =
  do D012._Token int D012._Natural
     D012._Token int D012._Natural
     D012._Token int D012._Natural
 
D012.Main : Grammar {} =
  do D012._RGB
     ($$ : {}) <- END
     pure $$
 
D012._Main : Grammar {} =
  do D012._RGB
     END
